In [2]:
%load_ext sql

In [3]:
%sql postgresql://postgres:root@localhost/sqlda 

In [4]:
pg_version=%sql select version()
print(pg_version)

 * postgresql://postgres:***@localhost/sqlda
1 rows affected.
+------------------------------------------------------------+
|                          version                           |
+------------------------------------------------------------+
| PostgreSQL 12.4, compiled by Visual C++ build 1914, 64-bit |
+------------------------------------------------------------+


In [5]:
%sql \dt

 * postgresql://postgres:***@localhost/sqlda
31 rows affected.


[('public', 'bat_emails', 'table', 'postgres'),
 ('public', 'bat_emails_threewks', 'table', 'postgres'),
 ('public', 'bat_ltd_sales', 'table', 'postgres'),
 ('public', 'bat_ltd_sales_count', 'table', 'postgres'),
 ('public', 'bat_ltd_sales_delay', 'table', 'postgres'),
 ('public', 'bat_ltd_sales_growth', 'table', 'postgres'),
 ('public', 'bat_ltd_sales_vol', 'table', 'postgres'),
 ('public', 'bat_sales', 'table', 'postgres'),
 ('public', 'bat_sales_daily', 'table', 'postgres'),
 ('public', 'bat_sales_daily_delay', 'table', 'postgres'),
 ('public', 'bat_sales_delay_vol', 'table', 'postgres'),
 ('public', 'bat_sales_growth', 'table', 'postgres'),
 ('public', 'closest_dealerships', 'table', 'postgres'),
 ('public', 'countries', 'table', 'postgres'),
 ('public', 'customer_sales', 'table', 'postgres'),
 ('public', 'customer_survey', 'table', 'postgres'),
 ('public', 'customers', 'table', 'postgres'),
 ('public', 'dealerships', 'table', 'postgres'),
 ('public', 'emails', 'table', 'postgres'),
 ('public', 'lemon_sales', 'table', 'postgres'),
 ('public', 'lemon_sales_count', 'table', 'postgres'),
 ('public', 'lemon_sales_delay', 'table', 'postgres'),
 ('public', 'lemon_sales_growth', 'table', 'postgres'),
 ('public', 'lemon_sales_sum', 'table', 'postgres'),
 ('public', 'product_names', 'table', 'postgres'),
 ('public', 'products', 'table', 'postgres'),
 ('public', 'products_sales', 'table', 'postgres'),
 ('public', 'public_transportation_by_zip', 'table', 'postgres'),
 ('public', 'sales', 'table', 'postgres'),
 ('public', 'salespeople', 'table', 'postgres'),
 ('public', 'top_cities_data', 'table', 'postgres')]

#### Analyzing Sales Growth by Email Opening Rate
To investigate the hypothesis that a decrease in the rate of opening emails impacted
the Bat Scooter sales rate, we will again select the Bat and Lemon Scooters and will
compare the email opening rate.

In [6]:
%%sql
SELECT * FROM emails LIMIT 5;

 * postgresql://postgres:***@localhost/sqlda
5 rows affected.


[(1, 18, 'Introducing A Limited Edition', 'f', 'f', 'f', datetime.datetime(2011, 1, 3, 15, 0), None, None),
 (2, 30, 'Introducing A Limited Edition', 'f', 'f', 'f', datetime.datetime(2011, 1, 3, 15, 0), None, None),
 (3, 41, 'Introducing A Limited Edition', 't', 'f', 'f', datetime.datetime(2011, 1, 3, 15, 0), datetime.datetime(2011, 1, 4, 10, 41, 11), None),
 (4, 52, 'Introducing A Limited Edition', 'f', 'f', 'f', datetime.datetime(2011, 1, 3, 15, 0), None, None),
 (5, 59, 'Introducing A Limited Edition', 'f', 'f', 'f', datetime.datetime(2011, 1, 3, 15, 0), None, None)]

* To investigate our hypothesis, we need to know whether an email was opened, and
when it was opened, as well as who the customer was who opened the email and
whether that customer purchased a scooter. 
* If the email marketing campaign was successful in maintaining the sales growth rate, we would expect a customer to open an email soon before a scooter was purchased.
* The period in which the emails were sent, as well as the ID of customers who
received and opened an email, can help us to determine whether a customer who
made a sale may have been encouraged to do so following the receipt of an email.
* To determine the hypothesis, we need to collect the **customer_id** column from
both the emails table and the bat_sales table for the Bat Scooter, the **opened,
sent_date, opened_date, and email_subject** columns from **emails** table, as well as
the **sales_transaction_date** column from the **bat_sales** table. 
* As we only want the email records of customers who purchased a Bat Scooter we will join the **customer_id** column in both tables. 
Then, insert the results into a new table – **bat_emails**

##### Email And Sales Information of the Lemon Scooter

In [8]:
%%sql
SELECT emails.email_subject, emails.customer_id, emails.opened,
emails.sent_date, emails.opened_date, bat_sales.sales_transaction_date
INTO bat_emails 
FROM emails 
INNER JOIN bat_sales 
ON bat_sales.customer_id=emails.customer_id 
ORDER BY bat_sales.sales_transaction_date;

 * postgresql://postgres:***@localhost/sqlda
40190 rows affected.


[]

In [10]:
%%sql
SELECT * FROM bat_emails LIMIT 5;

 * postgresql://postgres:***@localhost/sqlda
5 rows affected.


[('Like a Bat out of Heaven', 4553, 't', datetime.datetime(2016, 9, 21, 15, 0), datetime.datetime(2016, 9, 22, 9, 20, 41), datetime.datetime(2016, 10, 10, 0, 0)),
 ('A Brand New Scooter...and Car', 31307, 't', datetime.datetime(2014, 5, 6, 15, 0), datetime.datetime(2014, 5, 7, 9, 56, 56), datetime.datetime(2016, 10, 10, 0, 0)),
 ('Save the Planet with some Holiday Savings.', 42213, 't', datetime.datetime(2018, 11, 23, 15, 0), datetime.datetime(2018, 11, 24, 11, 57, 37), datetime.datetime(2016, 10, 10, 0, 0)),
 ('We Really Outdid Ourselves this Year', 24125, 'f', datetime.datetime(2017, 1, 15, 15, 0), None, datetime.datetime(2016, 10, 10, 0, 0)),
 ('Like a Bat out of Heaven', 40250, 'f', datetime.datetime(2016, 9, 21, 15, 0), None, datetime.datetime(2016, 10, 10, 0, 0))]

We can see here that there are several emails unopened, over a range of sent
dates, and that some customers have received multiple emails. Looking at the
subjects of the emails, some of them don't seem related to the Zoom scooters at
all.

##### Emails sent to customers before the sale transaction date

In [12]:
%%sql
SELECT * FROM bat_emails WHERE sent_date < sales_transaction_date
ORDER BY customer_id LIMIT 22;

 * postgresql://postgres:***@localhost/sqlda
22 rows affected.


[('Save the Planet with some Holiday Savings.', 7, 'f', datetime.datetime(2018, 11, 23, 15, 0), None, datetime.datetime(2019, 4, 25, 0, 0)),
 ('We cut you a deal: 20%% off a Blade', 7, 't', datetime.datetime(2014, 9, 18, 15, 0), datetime.datetime(2014, 9, 19, 15, 11, 17), datetime.datetime(2019, 4, 25, 0, 0)),
 ("Tis' the Season for Savings", 7, 'f', datetime.datetime(2015, 11, 26, 15, 0), None, datetime.datetime(2019, 4, 25, 0, 0)),
 ('Shocking Holiday Savings On Electric Scooters', 7, 'f', datetime.datetime(2013, 11, 29, 15, 0), None, datetime.datetime(2019, 4, 25, 0, 0)),
 ('Black Friday. Green Cars.', 7, 'f', datetime.datetime(2017, 11, 24, 15, 0), None, datetime.datetime(2019, 4, 25, 0, 0)),
 ("25% off all EVs. It's a Christmas Miracle!", 7, 't', datetime.datetime(2016, 11, 25, 15, 0), datetime.datetime(2016, 11, 26, 3, 55, 30), datetime.datetime(2019, 4, 25, 0, 0)),
 ('An Electric Car for a New Age', 7, 't', datetime.datetime(2015, 4, 1, 15, 0), datetime.datetime(2015, 4, 2, 15, 10, 55), datetime.datetime(2019, 4, 25, 0, 0)),
 ('A New Year, And Some New EVs', 7, 'f', datetime.datetime(2019, 1, 7, 15, 0), None, datetime.datetime(2019, 4, 25, 0, 0)),
 ('The 2013 Lemon Scooter is Here', 7, 'f', datetime.datetime(2013, 3, 1, 15, 0), None, datetime.datetime(2019, 4, 25, 0, 0)),
 ('Zoom Zoom Black Friday Sale', 7, 'f', datetime.datetime(2014, 11, 28, 15, 0), None, datetime.datetime(2019, 4, 25, 0, 0)),
 ('We Really Outdid Ourselves this Year', 7, 'f', datetime.datetime(2017, 1, 15, 15, 0), None, datetime.datetime(2019, 4, 25, 0, 0)),
 ('Like a Bat out of Heaven', 7, 'f', datetime.datetime(2016, 9, 21, 15, 0), None, datetime.datetime(2019, 4, 25, 0, 0)),
 ('A Brand New Scooter...and Car', 7, 'f', datetime.datetime(2014, 5, 6, 15, 0), None, datetime.datetime(2019, 4, 25, 0, 0)),
 ('A Brand New Scooter...and Car', 22, 't', datetime.datetime(2014, 5, 6, 15, 0), datetime.datetime(2014, 5, 7, 13, 31, 23), datetime.datetime(2017, 8, 14, 0, 0)),
 ('Shocking Holiday Savings On Electric Scooters', 22, 'f', datetime.datetime(2013, 11, 29, 15, 0), None, datetime.datetime(2017, 8, 14, 0, 0)),
 ('We cut you a deal: 20%% off a Blade', 22, 'f', datetime.datetime(2014, 9, 18, 15, 0), None, datetime.datetime(2017, 8, 14, 0, 0)),
 ('The 2013 Lemon Scooter is Here', 22, 'f', datetime.datetime(2013, 3, 1, 15, 0), None, datetime.datetime(2017, 8, 14, 0, 0)),
 ('An Electric Car for a New Age', 22, 'f', datetime.datetime(2015, 4, 1, 15, 0), None, datetime.datetime(2017, 8, 14, 0, 0)),
 ('Zoom Zoom Black Friday Sale', 22, 't', datetime.datetime(2014, 11, 28, 15, 0), datetime.datetime(2014, 11, 29, 11, 31, 3), datetime.datetime(2017, 8, 14, 0, 0)),
 ('Like a Bat out of Heaven', 22, 'f', datetime.datetime(2016, 9, 21, 15, 0), None, datetime.datetime(2017, 8, 14, 0, 0)),
 ('We Really Outdid Ourselves this Year', 22, 'f', datetime.datetime(2017, 1, 15, 15, 0), None, datetime.datetime(2017, 8, 14, 0, 0)),
 ("25% off all EVs. It's a Christmas Miracle!", 22, 'f', datetime.datetime(2016, 11, 25, 15, 0), None, datetime.datetime(2017, 8, 14, 0, 0))]

##### Remove unwanted email records, sent before the Bat Scooter was in production
Delete the rows of the bat_emails table where emails were sent more than **6
months prior to production**. 
As we can see, there are some emails that were sent
years before the transaction date. We can easily remove some of the unwanted
emails by removing those sent before the Bat Scooter was in production.
The technique used here is of efficiency and reduced storage key, instead of a traceable record of analysis used in previous study of Sales Price Hypothesis.

In [14]:
%%sql
DELETE FROM bat_emails WHERE sent_date < '2016-04-10';

 * postgresql://postgres:***@localhost/sqlda
12423 rows affected.


[]

##### Delete the rows where the sent date is after the purchase date, as they are not relevant to the sale

In [16]:
%%sql
DELETE FROM bat_emails WHERE sent_date > sales_transaction_date;

 * postgresql://postgres:***@localhost/sqlda
16112 rows affected.


[]

##### Emails sent close to the date of sale
* Delete those rows where the difference between the transaction date and the sent
date exceeds 30, as we also only want those emails that were sent shortly before
the scooter purchase. 
* An email 1 year beforehand is probably unlikely to influence
a purchasing decision, but one closer to the purchase date may have influenced
the sales decision. 
* We will set a limit of **1 month (30 days)** before the purchase.

In [19]:
%%sql
DELETE FROM bat_emails WHERE (sales_transaction_date-sent_date) >
'30 days';

 * postgresql://postgres:***@localhost/sqlda
10778 rows affected.


[]

In [21]:
%%sql
SELECT * FROM bat_emails ORDER BY customer_id LIMIT 22;

 * postgresql://postgres:***@localhost/sqlda
22 rows affected.


[("25% off all EVs. It's a Christmas Miracle!", 129, 't', datetime.datetime(2016, 11, 25, 15, 0), datetime.datetime(2016, 11, 26, 6, 31, 37), datetime.datetime(2016, 11, 28, 0, 0)),
 ('A New Year, And Some New EVs', 145, 'f', datetime.datetime(2019, 1, 7, 15, 0), None, datetime.datetime(2019, 1, 20, 0, 0)),
 ('Black Friday. Green Cars.', 150, 'f', datetime.datetime(2017, 11, 24, 15, 0), None, datetime.datetime(2017, 12, 19, 0, 0)),
 ('Black Friday. Green Cars.', 173, 'f', datetime.datetime(2017, 11, 24, 15, 0), None, datetime.datetime(2017, 12, 5, 0, 0)),
 ('We Really Outdid Ourselves this Year', 196, 'f', datetime.datetime(2017, 1, 15, 15, 0), None, datetime.datetime(2017, 1, 23, 0, 0)),
 ('We Really Outdid Ourselves this Year', 319, 'f', datetime.datetime(2017, 1, 15, 15, 0), None, datetime.datetime(2017, 1, 29, 0, 0)),
 ('Like a Bat out of Heaven', 369, 'f', datetime.datetime(2016, 9, 21, 15, 0), None, datetime.datetime(2016, 10, 13, 0, 0)),
 ('Like a Bat out of Heaven', 414, 'f', datetime.datetime(2016, 9, 21, 15, 0), None, datetime.datetime(2016, 10, 20, 0, 0)),
 ("25% off all EVs. It's a Christmas Miracle!", 418, 'f', datetime.datetime(2016, 11, 25, 15, 0), None, datetime.datetime(2016, 12, 21, 0, 0)),
 ('A New Year, And Some New EVs', 560, 't', datetime.datetime(2019, 1, 7, 15, 0), datetime.datetime(2019, 1, 8, 15, 56, 14), datetime.datetime(2019, 1, 29, 0, 0)),
 ('We Really Outdid Ourselves this Year', 600, 'f', datetime.datetime(2017, 1, 15, 15, 0), None, datetime.datetime(2017, 1, 18, 0, 0)),
 ('A New Year, And Some New EVs', 660, 't', datetime.datetime(2019, 1, 7, 15, 0), datetime.datetime(2019, 1, 8, 23, 37, 3), datetime.datetime(2019, 1, 8, 0, 0)),
 ('A New Year, And Some New EVs', 681, 'f', datetime.datetime(2019, 1, 7, 15, 0), None, datetime.datetime(2019, 1, 13, 0, 0)),
 ('Black Friday. Green Cars.', 806, 't', datetime.datetime(2017, 11, 24, 15, 0), datetime.datetime(2017, 11, 25, 16, 59, 40), datetime.datetime(2017, 11, 29, 0, 0)),
 ('A New Year, And Some New EVs', 881, 't', datetime.datetime(2019, 1, 7, 15, 0), datetime.datetime(2019, 1, 8, 21, 7, 28), datetime.datetime(2019, 1, 22, 0, 0)),
 ("25% off all EVs. It's a Christmas Miracle!", 934, 't', datetime.datetime(2016, 11, 25, 15, 0), datetime.datetime(2016, 11, 26, 9, 22, 45), datetime.datetime(2016, 12, 24, 0, 0)),
 ("25% off all EVs. It's a Christmas Miracle!", 983, 'f', datetime.datetime(2016, 11, 25, 15, 0), None, datetime.datetime(2016, 11, 29, 0, 0)),
 ('A New Year, And Some New EVs', 1060, 'f', datetime.datetime(2019, 1, 7, 15, 0), None, datetime.datetime(2019, 1, 27, 0, 0)),
 ("25% off all EVs. It's a Christmas Miracle!", 1288, 'f', datetime.datetime(2016, 11, 25, 15, 0), None, datetime.datetime(2016, 12, 11, 0, 0)),
 ("25% off all EVs. It's a Christmas Miracle!", 1317, 'f', datetime.datetime(2016, 11, 25, 15, 0), None, datetime.datetime(2016, 12, 13, 0, 0)),
 ('A New Year, And Some New EVs', 1400, 't', datetime.datetime(2019, 1, 7, 15, 0), datetime.datetime(2019, 1, 8, 15, 1), datetime.datetime(2019, 1, 10, 0, 0)),
 ('Save the Planet with some Holiday Savings.', 1417, 'f', datetime.datetime(2018, 11, 23, 15, 0), None, datetime.datetime(2018, 11, 26, 0, 0))]

We have reasonably filtered the available data based on the dates the
email was sent and opened. Looking at the preceding email_subject column, it also
appears that there are a few emails unrelated to the Bat Scooter, for example, **25%**
of all Electric Vehicle's. 

It's a Christmas Miracle! and Black Friday. Green Cars. These emails
seem more related to electric car production instead of scooters, and so we can
remove them from our analysis.

##### Filtering email's by relevance of subject
* Unique email subjects sent to potential customers of the Bat Scooter

In [23]:
%%sql
SELECT DISTINCT(email_subject) FROM bat_emails;

 * postgresql://postgres:***@localhost/sqlda
6 rows affected.


[('Black Friday. Green Cars.',),
 ("25% off all EVs. It's a Christmas Miracle!",),
 ('A New Year, And Some New EVs',),
 ('Like a Bat out of Heaven',),
 ('Save the Planet with some Holiday Savings.',),
 ('We Really Outdid Ourselves this Year',)]

* Delete all records that have Black Friday in the email subject. These emails do not
appear relevant to the sale of the Bat Scooter

In [25]:
%%sql
DELETE FROM bat_emails WHERE position('Black Friday' in email_subject)>0;

 * postgresql://postgres:***@localhost/sqlda
136 rows affected.


[]

* Delete all rows where 25% off all EVs. It's a Christmas Miracle! and A New Year,And Some New EVs can be found in the email_subject

In [28]:
%%sql
DELETE FROM bat_emails WHERE position('25% off all EV' in email_subject)>0;
DELETE FROM bat_emails WHERE position('Some New EV' in email_subject)>0;

 * postgresql://postgres:***@localhost/sqlda
0 rows affected.
199 rows affected.


[]

##### Final Dataset of emails sent to customers

Count of the final Bat scooter email dataset

In [30]:
%%sql
SELECT count(sales_transaction_date) FROM bat_emails;

 * postgresql://postgres:***@localhost/sqlda
1 rows affected.


[(401,)]

##### Count of opened Bat Scooter campaign emails

In [31]:
%%sql
SELECT count(opened) FROM bat_emails WHERE opened='t'

 * postgresql://postgres:***@localhost/sqlda
1 rows affected.


[(98,)]

##### Count of unique customers who receieved a Bat Scooter campaign email

In [32]:
%%sql
SELECT COUNT(DISTINCT(customer_id)) FROM bat_emails;

 * postgresql://postgres:***@localhost/sqlda
1 rows affected.


[(396,)]

##### Count of unique customers

In [34]:
%%sql
SELECT COUNT(DISTINCT(customer_id)) FROM bat_sales;

 * postgresql://postgres:***@localhost/sqlda
1 rows affected.


[(6659,)]

##### Percentage of customers who received an email

In [37]:
%%sql
SELECT (396.0/6659.0)*100 AS email_rate;

 * postgresql://postgres:***@localhost/sqlda
1 rows affected.


[(Decimal('5.94683886469439855800'),)]

In the preceding calculation, you can see that we included a decimal place in the
figures, for example, 396.0 instead of a simple integer value (396). This is because
the resulting value will be represented as less than 1 percentage point. If we
excluded these decimal places, the SQL server would have completed the division
operation as integers and the result would be 0.

Just under **6%** of customers who made a purchase received an email regarding
the Bat Scooter. Since **18%** of customers who received an email made a purchase,
there is a strong argument to be made that actively increasing the size of the
customer base who receive marketing emails could increase Bat Scooter sales.

##### Limit the email opening rate for the first 3 weeks, where there was a reduction in sales
Limit the scope of our data to be all sales prior to **November 1, 2016** and put the
data in a new table called **bat_emails_threewks**. So far, we have examined the email
opening rate throughout all available data for the Bat Scooter. Check the rate
throughout for the first 3 weeks, where we saw a reduction in sales

In [38]:
%%sql
SELECT * 
INTO bat_emails_threewks 
FROM bat_emails 
WHERE sales_transaction_date < '2016-11-01';

 * postgresql://postgres:***@localhost/sqlda
82 rows affected.


[]

* Count the number of emails sent in the first 3 weeks
* We can see that we have sent **82** emails during this period

In [39]:
%%sql
SELECT COUNT(opened) FROM bat_emails_threewks;

 * postgresql://postgres:***@localhost/sqlda
1 rows affected.


[(82,)]

* Count the number of emails opened in the first 3 weeks
* We can see that **15** emails were opened in the first 3 weeks. 

In [41]:
%%sql
SELECT COUNT(opened) FROM bat_emails_threewks WHERE opened='t';

 * postgresql://postgres:***@localhost/sqlda
1 rows affected.


[(15,)]

* Count the number of customers who received emails during the first 3 weeks of
sales and who then made a purchase

In [43]:
%%sql
SELECT COUNT(DISTINCT(customer_id)) FROM bat_emails_threewks;

 * postgresql://postgres:***@localhost/sqlda
1 rows affected.


[(82,)]

* Calculate the percentage of customers who opened emails pertaining to the Bat
Scooter and then made a purchase in the first 3 weeks

In [45]:
%%sql
SELECT (15.0/82.0)*100 AS sale_rate;

 * postgresql://postgres:***@localhost/sqlda
1 rows affected.


[(Decimal('18.29268292682926829300'),)]

Approximately **18%** of customers who received an email about the Bat Scooter
made a purchase in the first 3 weeks. This is consistent with the rate for all
available data for the Bat Scooter.

##### Calculate how many unique customers we have in total throughout the first 3 weeks

This information is useful context when considering the percentages,
we just calculated. 3 sales out of 4 equate to 75% but, in this situation, we
would prefer a lower rate of the opening but for a much larger customer base.
Information on larger customer bases is generally more useful as it is typically
more representative of the entire customer base, rather than a small sample of it.
We already know that 82 customers received emails

* Number of distinct customers from bat_sales

In [7]:
%%sql
SELECT COUNT(DISTINCT(customer_id)) FROM bat_sales WHERE sales_transaction_date < '2016-11-01';

 * postgresql://postgres:***@localhost/sqlda
1 rows affected.


[(160,)]

* The following output reflects 160 customers where the transaction took place
before November 1, 2016. 
* There were **160** customers in the first 3 weeks, **82** of whom received emails, which
is slightly over **50%** of customers. This is much more than **6%** of customers over
the entire period of availability of the scooter.
* Now that we have examined the performance of the email marketing campaign for the
Bat Scooter, we need a **control or comparison group** to establish whether the results
were consistent with that of other products. Without a group to compare against, we
simply do not know whether the email campaign of the Bat Scooter was good, bad, or
neither.